packages/react/dialog/src/Dialog.tsx

```diff
@@ -247,31 +247,41 @@ const DialogContentModal = React.forwardRef((props, forwardedRef) => {
 
 const DialogContentNonModal = React.forwardRef((props, forwardedRef) => {
   const context = useDialogContext(CONTENT_NAME);
-  const isEscapeKeyDownRef = React.useRef(false);
+  const isInteractOutsideRef = React.useRef(false);
+
   return (
     <Portal>
       <DialogContentImpl
         {...props}
         ref={forwardedRef}
         trapFocus={false}
         disableOutsidePointerEvents={false}
-        onCloseAutoFocus={composeEventHandlers(props.onCloseAutoFocus, (event) => {
-          event.preventDefault();
-          if (isEscapeKeyDownRef.current) context.triggerRef.current?.focus();
-        })}
-        onEscapeKeyDown={composeEventHandlers(props.onEscapeKeyDown, () => {
-          isEscapeKeyDownRef.current = true;
-        })}
-        onInteractOutside={composeEventHandlers(props.onInteractOutside, (event) => {
+        onCloseAutoFocus={(event) => {
+          props.onCloseAutoFocus?.(event);
+          const userPrevented = event.defaultPrevented;
+
+          if (!userPrevented) {
+            event.preventDefault();
+            if (!isInteractOutsideRef.current) context.triggerRef.current?.focus();
+          }
+
+          isInteractOutsideRef.current = false;
+        }}
+        onInteractOutside={(event) => {
+          props.onInteractOutside?.(event);
+          const userPrevented = event.defaultPrevented;
+
           // Prevent dismissing when clicking the trigger.
           // As the trigger is already setup to close, without doing so would
           // cause it to close and immediately open.
           //
           // We use `onInteractOutside` as some browsers also
           // focus on pointer down, creating the same issue.
-          const targetIsTrigger = context.triggerRef.current?.contains(event.target as HTMLElement);
+          const target = event.target as HTMLElement;
+          const targetIsTrigger = context.triggerRef.current?.contains(target);
           if (targetIsTrigger) event.preventDefault();
-        })}
+          else if (!userPrevented) isInteractOutsideRef.current = true;
```

Comment:
again, I'm fine with just `else if (!event.defaultPrevented)` here because it's not necessarily the user preventing it.

Reply:
Yeah I'm on board with this 👍